python/opae.admin/opae/admin/tools/fpgasupdate.py

```diff
@@ -351,27 +381,183 @@ def canonicalize_bdf(bdf):
+
+
+def fpga_err_to_string(err_code):
+    if err_code == FPGA_IMAGE_ERR_HW_ERROR:
+        return "hw_error"
+    if err_code == FPGA_IMAGE_ERR_TIMEOUT:
+        return "timeout"
+    if err_code == FPGA_IMAGE_ERR_CANCELED:
+        return "cancelled"
+    if err_code == FPGA_IMAGE_ERR_BUSY:
+        return "busy"
+    if err_code == FPGA_IMAGE_ERR_INVALID_SIZE:
+        return "invalid size"
+    if err_code == FPGA_IMAGE_ERR_RW_ERROR:
+        return "rw_error"
+    if err_code == FPGA_IMAGE_ERR_WEAROUT:
+        return "flash_wearout"
+    return ""
+
+
 class SecureUpdateError(Exception):
     """Secure update exception"""
     def __init__(self, arg):
         super(SecureUpdateError, self).__init__(self)
         self.errno, self.strerror = arg
 
 
-def update_fw(args, pac):
+def update_fw_ioctl(sec_dev, infile, pac):
     """Writes firmware to secure device.
 
-    args - the object resulting from command-line parsing.
+    sec_dev - an integer file descriptor to the os.open()'ed secure device file.
+    infile - the image to write to the FPGA
+    pac - opae fpga module
 
     returns a 2-tuple of the process exit status and a message.
     """
-    # bytes/sec when staging is flash
-    flash_copy_bps = 43000.0
-    # bytes/sec when staging area is dram
-    dram_copy_bps = 92000.0
-    dram_copy_offset = 42.0
 
-    infile = args.file
+    orig_pos = infile.tell()
+    infile.seek(0, os.SEEK_END)
+    payload_size = infile.tell() - orig_pos
+
+    LOG.info('updating from file %s with size %d',
+            infile.name, payload_size)
+
+    file_buf = array.array('B')
+    infile.seek(orig_pos, os.SEEK_SET)
+    file_buf.fromfile(infile, payload_size)
+    infile.close() # TODO: Is this the right place to call .close()?
+    addr, _ = file_buf.buffer_info()
+
+    # Create the eventfd file descriptor
+    clib = find_library("c")
+    if clib is None:
+        return errno.EINVAL, 'C library not found'
+    libc = cdll.LoadLibrary(clib)
+    evtfd = libc.eventfd(0, 0)
+
+    write_header = array.array('B', [0] * 32)
+    #  struct fpga_image_write {
+    # 	/* Input */
+    # 	__u32 flags;		/* Zero for now */
+    # 	__u32 size;		/* Data size (in bytes) to be written */
+    # 	__s32 evtfd;		/* File descriptor for completion signal */
+    # 	__u64 buf;		/* User space address of source data */
+    # };
+    struct.pack_into('IIIQ', write_header, 0, 0, payload_size, evtfd, addr)
+
+    LOG.info('waiting for idle')
+    retries = 0
+    timeout = 1.0
+    max_retries = 60 * 5
+    _, progress_code, _, _ = ioctl_status(sec_dev)
+    while progress_code != FPGA_IMAGE_PROG_IDLE:
+        time.sleep(timeout)
+        retries += 1
+        if retries > max_retries:
+            return errno.ETIMEDOUT, 'Secure update timed out'
+        _, progress_code, _, _ = ioctl_status(sec_dev)
+
+    # IOCTL Call to kick off the write
+    fcntl.ioctl(sec_dev, IOCTL_FPGA_IMAGE_LOAD_WRITE, write_header, True)
```

Comment:
Need a try: here to guard against ioctl exceptions. If the ioctl() fails, an OSError exception is raised.